### ISSUE TITLE
Work around Clang's inconsistent traversal of desugared lambdas and add test cases [phab:D1904]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2346]
_Created by **shahms** at 2017-08-23 22:54:28_: 

Clang only traverses generated lambda code at namespace scope, not nested scopes.
As a result, the most comment cases of lambda use are undecorated.  Fix this with an ugly
work around until we can revisit this more completely.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-08-24 17:11:07_:

Why a thunk and not a conditional expression?
```
RecursiveASTVisitor::TraverseLambdaExpr(Expr) &&
  (Expr->getLambdaClass()->getDeclContext()->isFunctionOrMethod() ?
    TraverseDecl(Expr->getCallOperator()) : true);
```

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2017-08-24 17:11:07_

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2017-08-24 17:14:36_:

I thought the thunk was easier to read? I could have also gone with nested ifs, which might have been better.

**Comment by kythe** [comment]
_**shahms** closed via commit f75affeed3aa5ef3f7d4e50177dac394cb2d2903 at 2017-08-24 17:15:32_
